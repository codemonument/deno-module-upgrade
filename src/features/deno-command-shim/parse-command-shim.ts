import { CommandShim } from "./command-shim.type.ts";
import { ExecCommand } from "./exec-command.type.ts";
import { parseExecCommand } from "./parse-exec-command.ts";
import { DenoModuleUrlSegments } from "../deno-module-registries/deno-module-url-segments.type.ts";
import { DenoModuleRegistry } from "../deno-module-registries/deno-module-registry.type.ts";
import { detectModuleRegistry } from "../deno-module-registries/detect-module-registry.ts";

export interface ParseCommandShimOptions {
  /**
   * Default: true
   * If set to false, parsing function will not validate,
   * whether it was passed a deno shim or not.
   */
  validateDenoShim: boolean;
}

const defaultOptions: ParseCommandShimOptions = {
  validateDenoShim: true,
};

/**
 * @param shimPath the path of the command shim
 * @param shimContent the content of the command shim as string
 * (so that this parser does not need to load the file itself - better for testing)
 */
export function parseCommandShim(
  shimPath: string,
  shimContent: string | string[],
  options: Partial<ParseCommandShimOptions> = {},
): CommandShim {
  const { validateDenoShim } = { ...defaultOptions, ...options };

  const shimLines = (typeof shimContent === "string")
    ? shimContent.split("\n")
    : shimContent;
  const [_shebang, denoComment, commandString] = shimLines;

  if (validateDenoShim && denoComment !== "# generated by deno install") {
    throw new Deno.errors.InvalidData(
      `Command Shim "${shimPath}" may not have been generated by "deno install"! 
      Force proceed by running again with --validateDenoShim=false. 
    
      File Content: 
 
      ${shimContent}`,
    );
  }

  const execCommand: ExecCommand = parseExecCommand(commandString);
  const moduleRegistry: DenoModuleRegistry = detectModuleRegistry(
    execCommand.moduleURL,
  );

  const moduleUrlSegments: DenoModuleUrlSegments = moduleRegistry
    .parseModuleUrlSegments(execCommand.moduleURL);

  return {
    path: shimPath,
    lines: shimLines,
    commandString,
    moduleRegistry,
    execCommand,
    moduleUrlSegments,
  };
}
