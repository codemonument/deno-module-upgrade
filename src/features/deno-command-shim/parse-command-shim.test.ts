import {
  assertSnapshot,
  assertThrows,
  describe,
  it,
} from "../../deps/testing.std.ts";
import { parseCommandShim } from "./parse-command-shim.ts";

describe(`Function parse-deno-command-shim`, () => {
  describe(`with validateDenoShim === true`, () => {
    it(`should return correct shim datastructure`, (t) => {
      const shimLines = [
        "#!/bin/sh",
        "# generated by deno install",
        `exec deno run --allow-read=deps.ts --allow-write=deps.ts --allow-net=cdn.deno.land,api.deno.land,x.nest.land,raw.githubusercontent.com,github.com,api.github.com 'https://x.nest.land/dmm@2.1.0/mod.ts' "$@"`,
      ];
      const res = parseCommandShim("fakePath", shimLines);

      assertSnapshot(t, res);
    });

    it(`should throw on wrong deno comment`, () => {
      assertThrows(() => {
        parseCommandShim("fakePath", [
          "#!/bin/sh",
          "# NOT generated by deno install - fake comment",
          `exec deno run --allow-read=deps.ts --allow-write=deps.ts --allow-net=cdn.deno.land,api.deno.land,x.nest.land,raw.githubusercontent.com,github.com,api.github.com 'https://x.nest.land/dmm@2.1.0/mod.ts' "$@"`,
        ]);
      });
    });
  });
});
