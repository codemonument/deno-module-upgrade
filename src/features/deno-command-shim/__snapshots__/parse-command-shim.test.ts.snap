export const snapshot = {};

snapshot[`Function parse-deno-command-shim > with validateDenoShim === true > should return correct shim datastructure (with deno.land) 1`] = `
{
  commandString: \`exec deno run 'https://deno.land/x/cowsay/cowsay.ts' "\$@"\`,
  execCommand: {
    execCommand: \`exec deno run 'https://deno.land/x/cowsay/cowsay.ts' "\$@"\`,
    execFlags: [
    ],
    moduleURL: URL {
  href: "https://deno.land/x/cowsay/cowsay.ts",
  origin: "https://deno.land",
  protocol: "https:",
  username: "",
  password: "",
  host: "deno.land",
  hostname: "deno.land",
  port: "",
  pathname: "/x/cowsay/cowsay.ts",
  hash: "",
  search: ""
},
  },
  lines: [
    "#!/bin/sh",
    "# generated by deno install",
    \`exec deno run 'https://deno.land/x/cowsay/cowsay.ts' "\$@"\`,
  ],
  moduleRegistry: DenoLandRegistry {},
  moduleUrlSegments: {
    completeModuleURL: URL {
  href: "https://deno.land/x/cowsay/cowsay.ts",
  origin: "https://deno.land",
  protocol: "https:",
  username: "",
  password: "",
  host: "deno.land",
  hostname: "deno.land",
  port: "",
  pathname: "/x/cowsay/cowsay.ts",
  hash: "",
  search: ""
},
    entryFileName: "cowsay.ts",
    entryFilePath: "cowsay.ts",
    moduleBaseURL: URL {
  href: "https://deno.land/x/cowsay",
  origin: "https://deno.land",
  protocol: "https:",
  username: "",
  password: "",
  host: "deno.land",
  hostname: "deno.land",
  port: "",
  pathname: "/x/cowsay",
  hash: "",
  search: ""
},
    moduleName: "cowsay",
    moduleVersion: undefined,
  },
  path: "fakePath",
}
`;

snapshot[`Function parse-deno-command-shim > with validateDenoShim === true > should throw on wrong deno comment 1`] = `
Error: Command Shim "fakePath" may not have been generated by "deno install"! 
           Force proceed by running again with --validateDenoShim=false. 
          
           File Content: 
      
           #!/bin/sh,# NOT generated by deno install - fake comment,exec deno run 'https://deno.land/x/cowsay/cowsay.ts' "\$@"
    at parseCommandShim (file:///Users/bjesuiter/Develop/codemonument/deno-modup/src/features/deno-command-shim/parse-command-shim.ts:42:7)
    at Object.<anonymous> (file:///Users/bjesuiter/Develop/codemonument/deno-modup/src/features/deno-command-shim/parse-command-shim.test.ts:25:19)
    at Function.runTest (https://deno.land/std@0.141.0/testing/_test_suite.ts:358:16)
    at Function.runTest (https://deno.land/std@0.141.0/testing/_test_suite.ts:346:33)
    at Function.runTest (https://deno.land/std@0.141.0/testing/_test_suite.ts:346:33)
    at fn (https://deno.land/std@0.141.0/testing/_test_suite.ts:316:37)
    at testStepSanitizer (deno:runtime/js/40_testing.js:449:13)
    at asyncOpSanitizer (deno:runtime/js/40_testing.js:149:15)
    at resourceSanitizer (deno:runtime/js/40_testing.js:375:13)
    at exitSanitizer (deno:runtime/js/40_testing.js:432:15)
`;

snapshot[`Function parse-deno-command-shim > with validateDenoShim === true > should parse nest.land shim  1`] = `
{
  commandString: \`exec deno run --allow-read=deps.ts --allow-write=deps.ts --allow-net=cdn.deno.land,api.deno.land,x.nest.land,raw.githubusercontent.com,github.com,api.github.com 'https://x.nest.land/dmm@2.1.0/mod.ts' "\$@"\`,
  execCommand: {
    execCommand: \`exec deno run --allow-read=deps.ts --allow-write=deps.ts --allow-net=cdn.deno.land,api.deno.land,x.nest.land,raw.githubusercontent.com,github.com,api.github.com 'https://x.nest.land/dmm@2.1.0/mod.ts' "\$@"\`,
    execFlags: [
      "--allow-read=deps.ts",
      "--allow-write=deps.ts",
      "--allow-net=cdn.deno.land,api.deno.land,x.nest.land,raw.githubusercontent.com,github.com,api.github.com",
    ],
    moduleURL: URL {
  href: "https://x.nest.land/dmm@2.1.0/mod.ts",
  origin: "https://x.nest.land",
  protocol: "https:",
  username: "",
  password: "",
  host: "x.nest.land",
  hostname: "x.nest.land",
  port: "",
  pathname: "/dmm@2.1.0/mod.ts",
  hash: "",
  search: ""
},
  },
  lines: [
    "#!/bin/sh",
    "# generated by deno install",
    \`exec deno run --allow-read=deps.ts --allow-write=deps.ts --allow-net=cdn.deno.land,api.deno.land,x.nest.land,raw.githubusercontent.com,github.com,api.github.com 'https://x.nest.land/dmm@2.1.0/mod.ts' "\$@"\`,
  ],
  moduleRegistry: NestLandRegistry {},
  moduleUrlSegments: {
    completeModuleURL: URL {
  href: "https://x.nest.land/dmm@2.1.0/mod.ts",
  origin: "https://x.nest.land",
  protocol: "https:",
  username: "",
  password: "",
  host: "x.nest.land",
  hostname: "x.nest.land",
  port: "",
  pathname: "/dmm@2.1.0/mod.ts",
  hash: "",
  search: ""
},
    entryFileName: "mod.ts",
    entryFilePath: "mod.ts",
    moduleBaseURL: URL {
  href: "https://x.nest.land/dmm",
  origin: "https://x.nest.land",
  protocol: "https:",
  username: "",
  password: "",
  host: "x.nest.land",
  hostname: "x.nest.land",
  port: "",
  pathname: "/dmm",
  hash: "",
  search: ""
},
    moduleName: "dmm",
    moduleVersion: "2.1.0",
  },
  path: "fakePath",
}
`;
